File: selenium-java/src/test/java/com/framework/tests/SanityTest.java
package com.framework.tests;

import com.framework.utils.ConfigReader;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.testng.Assert;
import org.testng.annotations.Test;

import java.net.HttpURLConnection;
import java.net.URL;
import java.time.Duration;

/**
 * SanityTest: Validates the automation infrastructure itself — not the app under test.
 *
 * <p>WHY SANITY TESTS EXIST: Before trusting test results, you must trust the test
 * environment. If Chrome is not installed, if the config file is missing, or if the
 * Selenium Grid is down, every test in the suite will fail for infrastructure reasons
 * rather than application defects. Sanity tests diagnose those infrastructure problems
 * directly, making it immediately clear whether failures are "the app is broken" or
 * "the test environment is broken."
 *
 * <p>WHY THIS CLASS DOES NOT EXTEND {@link com.framework.base.BaseTest}:
 * This is intentional and important. BaseTest uses DriverFactory, DriverManager, and
 * the full framework stack. If any part of that stack is broken, BaseTest itself would
 * fail — which would mask the real cause. SanityTest bypasses the framework entirely,
 * using raw Selenium directly. If SanityTest passes but other tests fail, the problem
 * is in the framework, not the environment. If SanityTest fails, the environment itself
 * is broken. This isolation is what makes sanity tests valuable.
 *
 * <p>WHAT THIS CLASS COVERS:
 * <ol>
 *   <li><b>Browser engine</b> — "bare metal" ChromeDriver launch without any framework
 *       abstractions. Proves Chrome is installed and functional on this machine/CI node.</li>
 *   <li><b>Configuration integrity</b> — reads config.properties and confirms critical
 *       values (URL, timeout) are present and correctly formatted.</li>
 *   <li><b>Grid connectivity</b> — makes an HTTP ping to the Selenium Grid hub to confirm
 *       it is reachable before tests that depend on it are run.</li>
 * </ol>
 */
public class SanityTest {

    // ---------------------------------------------------------
    // HAPPY PATH (Sanity): Browser Engine Check
    // "Bare Metal" test - No DriverFactory, no listeners.
    // If this fails, the machine/CI node is broken.
    // ---------------------------------------------------------

    /**
     * testRawSeleniumLaunch: Launches ChromeDriver directly without any framework
     * abstractions and verifies the SauceDemo login page renders.
     *
     * <p>WHY "BARE METAL": This test intentionally avoids DriverFactory, DriverManager,
     * BaseTest, and all framework utilities. It uses only raw Selenium 4 APIs.
     * This is the most honest test of whether "Selenium works on this machine" —
     * there is no framework code between this test and the browser that could
     * hide or cause the failure.
     *
     * <p>WHY HEADLESS HERE: CI environments typically have no graphical display.
     * These Chrome options match what most CI pipelines require:
     * <ul>
     *   <li>{@code --headless=new} — headless mode without a window.</li>
     *   <li>{@code --no-sandbox} — required in Docker containers where the Linux
     *       user namespace sandbox is not available.</li>
     *   <li>{@code --disable-dev-shm-usage} — prevents /dev/shm (shared memory)
     *       from being too small in Docker, which causes Chrome to crash.</li>
     * </ul>
     *
     * <p>WHY THE LOGO ASSERTION: Rather than just checking the URL (which could
     * succeed while the page renders blank), this test waits for a visible UI element —
     * the login logo — confirming the page actually rendered meaningful content.
     */
    @Test(groups = {"sanity", "smoke", "web"})
    public void testRawSeleniumLaunch() {
        System.out.println("[INFO] Starting Bare-Metal Sanity Check...");
        WebDriver driver = null;

        try {
            // Configure Chrome options for headless CI execution.
            ChromeOptions options = new ChromeOptions();
            options.addArguments("--headless=new");
            options.addArguments("--no-sandbox");
            options.addArguments("--disable-dev-shm-usage");
            options.addArguments("--remote-allow-origins=*");

            // Selenium 4 includes Selenium Manager, which automatically downloads
            // the correct ChromeDriver version — no manual ChromeDriver setup needed.
            driver = new ChromeDriver(options);

            // Navigate using URL from config — keeps this test environment-agnostic.
            driver.get(ConfigReader.getProperty("url"));

            // Explicit wait: poll up to 5 seconds for the logo to appear.
            // This is more reliable than an implicit wait because it waits for a
            // specific, meaningful element — not just any element on the page.
            WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(5));
            WebElement logo = wait.until(ExpectedConditions.visibilityOfElementLocated(By.className("login_logo")));

            // Confirm the logo element is actually displayed (visible, not hidden).
            Assert.assertTrue(logo.isDisplayed(), "Raw Selenium failed to render the page!");
            System.out.println("[PASS] Browser Engine is operational.");

        } catch (Exception e) {
            // Any unexpected exception (ChromeDriver not found, page load timeout, etc.)
            // is wrapped in a clear failure message that points to the root cause.
            Assert.fail("CRITICAL: Selenium Environment is broken! " + e.getMessage());
        } finally {
            // Always quit the driver, even if the test failed.
            // Without this, the Chrome process would remain running indefinitely.
            if (driver != null) driver.quit();
        }
    }

    // ---------------------------------------------------------
    // EDGE CASE (Sanity): Configuration Integrity
    // Verifies that critical secrets/config are actually loaded.
    // ---------------------------------------------------------

    /**
     * testConfigurationLoad: Verifies that the config.properties file is present,
     * readable, and contains valid values for critical settings.
     *
     * <p>WHY TEST THE CONFIG: If config.properties is missing or corrupt, every test
     * that calls {@code ConfigReader.getProperty()} will get null back and fail with
     * cryptic NullPointerExceptions in page objects and assertions rather than a clear
     * "config is broken" message. This test provides that clear early warning.
     *
     * <p>WHAT IS VALIDATED:
     * <ul>
     *   <li>{@code url} — the base URL of the app under test. Missing = all navigation fails.</li>
     *   <li>{@code timeout.explicit} — the explicit wait timeout in seconds. Must be a
     *       parseable integer; a non-numeric value would crash all page objects that
     *       use explicit waits.</li>
     * </ul>
     */
    @Test(groups = {"sanity"})
    public void testConfigurationLoad() {
        System.out.println("[INFO] Checking Config Integrity...");

        // Verify the base URL is present and non-empty.
        // Null means the key doesn't exist in config; empty means it was set to nothing.
        // Both are fatal for any navigation-dependent test.
        String url = ConfigReader.getProperty("url");
        Assert.assertNotNull(url, "Config Failure: 'url' is missing in config.properties");
        Assert.assertFalse(url.isEmpty(), "Config Failure: 'url' is empty");

        // Verify the explicit timeout value is a valid integer.
        // parseInt() throws NumberFormatException if the value is e.g. "ten" or empty.
        // This assertion converts that exception into a clear, actionable test failure.
        String timeout = ConfigReader.getProperty("timeout.explicit");
        try {
            Integer.parseInt(timeout);
        } catch (NumberFormatException e) {
            Assert.fail("Config Failure: 'timeout.explicit' is not a valid number.");
        }

        System.out.println("[PASS] Configuration file is readable and valid.");
    }

    // ---------------------------------------------------------
    // INFRASTRUCTURE TEST (Sanity): Grid Connectivity
    // Checks if the Docker Grid (Hub) is actually up and listening.
    // ---------------------------------------------------------

    /**
     * testHubConnectivity: Checks whether the Selenium Grid hub is reachable
     * by sending a lightweight HTTP GET to its {@code /status} endpoint.
     *
     * <p>WHY CHECK THE GRID: Tests that use {@code -Dtarget=grid} (Docker Compose or
     * Kubernetes deployment) will hang or fail immediately if the Grid hub isn't running.
     * This sanity test catches that before any tests run, saving time diagnosing
     * "why did 50 tests all fail with connection refused?"
     *
     * <p>WHY NOT FAIL IF HUB IS UNREACHABLE: This test is expected to run in all
     * environments, including local development where Docker may not be running.
     * A hard {@code Assert.fail()} would break local runs where the Grid simply
     * isn't needed. Instead, the test logs a warning and passes — only in a dedicated
     * CI environment with Grid would you un-comment the hard failure assertion.
     *
     * <p>WHY 2-SECOND TIMEOUT: The Grid should respond in milliseconds if it's up.
     * A 2-second {@code connectTimeout} means we "fail fast" without hanging the
     * test suite if the hub is down. Long timeouts (the default is often 30+ seconds)
     * would make the sanity check annoyingly slow to fail.
     */
    @Test(groups = {"sanity", "docker"})
    public void testHubConnectivity() {
        String hubUrl = ConfigReader.getProperty("hub.url");

        // Log context about which hub we're checking, for CI debugging.
        if (hubUrl == null || hubUrl.contains("localhost")) {
            System.out.println("[INFO] Checking connectivity to Local Grid: " + hubUrl);
        }

        try {
            // Java's built-in HttpURLConnection is used here deliberately —
            // it's a lightweight HTTP ping with no external library dependencies,
            // which is appropriate for an infrastructure health check.
            URL url = new URL(hubUrl + "/status"); // Selenium Grid 4 status endpoint
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");
            connection.setConnectTimeout(2000); // Fail fast (2 seconds) if hub is down

            int responseCode = connection.getResponseCode();

            // HTTP 200 OK means the Selenium Grid hub is alive and accepting connections.
            Assert.assertEquals(responseCode, 200, "Selenium Grid is not responding! Response Code: " + responseCode);
            System.out.println("[PASS] Selenium Grid is reachable.");

        } catch (Exception e) {
            // Grid not running is expected in local dev — warn but don't fail.
            // In a dedicated CI environment that requires the Grid, uncomment
            // Assert.fail() below to turn this into a hard failure:
            // Assert.fail("Grid Infrastructure is DOWN.");
            System.out.println("[WARN] Grid check failed (This is expected if not running Docker): " + e.getMessage());
        }
    }
}

File: playwright/tests/Framework.Tests/Tests/SanityTest.cs
using Microsoft.Playwright;
using Framework.Core.Config;

namespace Framework.Tests.Tests;

/// <summary>
/// SanityTest: "Bare metal" infrastructure checks that validate the test environment
/// itself, not the application under test.
///
/// WHY SANITY TESTS EXIST: Before trusting test results, you must trust the test
/// environment. If Playwright can't launch a browser, if appsettings.json is missing
/// a required key, or if the Selenium Grid hub isn't running, every test will fail for
/// infrastructure reasons — not application bugs. Sanity tests diagnose these problems
/// directly, giving a clear "environment is broken" signal before any app tests run.
///
/// WHY THIS CLASS DOES NOT EXTEND BaseTest:
/// BaseTest contains shared setup and teardown logic. If BaseTest itself is broken,
/// tests extending it would all fail in setUp — masking the real cause. SanityTest
/// bypasses BaseTest entirely, using Playwright APIs directly. This isolation means:
///   - If SanityTest passes but other tests fail → the issue is in the framework code.
///   - If SanityTest fails → the environment itself is broken (browser, config, etc.).
///
/// MIRRORS THE JAVA SanityTest: This C# class follows the same design as the Java
/// framework's SanityTest — one of the ways the portfolio demonstrates cross-language
/// consistency in testing principles.
///
/// WHAT THIS CLASS COVERS:
///   1. RawPlaywrightLaunch  — creates a browser directly (no BaseTest wrappers)
///      to confirm Playwright and browsers are correctly installed.
///   2. ConfigurationLoad    — reads appsettings.json and confirms required keys
///      (url, username, password) are present and non-empty.
///   3. HubConnectivity      — pings the Selenium/Playwright Grid hub URL to confirm
///      it's reachable before grid-targeted tests run.
/// </summary>
[TestFixture]
public class SanityTest
{
    /// <summary>
    /// RawPlaywrightLaunch: Creates a Playwright browser instance manually (without
    /// BaseTest or PageTest) and navigates to SauceDemo to confirm the browser works.
    ///
    /// WHY "BARE METAL": This test intentionally avoids all framework abstractions —
    /// no BaseTest, no ConfigReader for the URL, no page objects. It uses only the raw
    /// Playwright.NET API. This is the most honest test of whether "Playwright works
    /// on this machine" because there is no framework code between the test and the
    /// browser that could hide or cause the failure.
    ///
    /// WHY HEADLESS: CI environments have no graphical display. Headless = true
    /// launches Chrome without a visible window, which works in any environment.
    ///
    /// RESOURCE MANAGEMENT: The test uses 'using' and 'await using' to ensure
    /// IPlaywright and IBrowser are disposed (closed/cleaned up) even if the
    /// test throws an exception — preventing orphan browser processes.
    /// </summary>
    [Test]
    [Category("sanity")]
    public async Task RawPlaywrightLaunch()
    {
        // Playwright.CreateAsync() is the entry point for the entire Playwright library.
        // It returns the top-level object from which you create browsers.
        using IPlaywright playwright = await Playwright.CreateAsync();

        // Launch Chromium (the open-source base of Google Chrome) in headless mode.
        // 'await using' ensures the browser is closed and resources freed after the test.
        await using IBrowser browser = await playwright.Chromium.LaunchAsync(
            new BrowserTypeLaunchOptions { Headless = true });

        // Open a new browser tab (page) to navigate to the application.
        IPage page = await browser.NewPageAsync();
        await page.GotoAsync("https://www.saucedemo.com");

        // Verify navigation succeeded — the URL should contain "saucedemo".
        // This confirms the browser launched, the network is reachable, and
        // the page responded (not a connection refused or DNS failure).
        Assert.That(page.Url, Does.Contain("saucedemo"),
            "Raw Playwright should navigate to saucedemo.com.");

        await browser.CloseAsync();
    }

    /// <summary>
    /// ConfigurationLoad: Reads appsettings.json and verifies that the three
    /// critical keys (url, app:username, app:password) are present and non-empty.
    ///
    /// WHY TEST THE CONFIG: Every test in the framework depends on these values.
    /// If appsettings.json is missing, malformed, or doesn't contain the expected keys,
    /// every other test will fail with cryptic NullReferenceExceptions or empty-string
    /// errors deep inside page objects. This test provides an early, clear failure
    /// message ("app:username must be set") rather than a confusing late failure.
    ///
    /// ASSERT.MULTIPLE: Collects all three assertions before reporting failures.
    /// If all three keys are missing, you see all three failures at once rather than
    /// fixing them one at a time — a more efficient debugging experience.
    /// </summary>
    [Test]
    [Category("sanity")]
    public void ConfigurationLoad()
    {
        // Read the three most critical configuration values.
        // ConfigReader reads from appsettings.json and supports environment variable
        // overrides — so CI pipelines can inject values without modifying the file.
        string url      = ConfigReader.GetProperty("url");
        string username = ConfigReader.GetProperty("app:username");
        string password = ConfigReader.GetProperty("app:password");

        // Collect all assertion failures before reporting — if multiple keys are
        // missing, all are reported in one test run rather than one-at-a-time.
        Assert.Multiple(() =>
        {
            Assert.That(url,      Is.Not.Empty, "url must be set in appsettings.json");
            Assert.That(username, Is.Not.Empty, "app:username must be set");
            Assert.That(password, Is.Not.Empty, "app:password must be set");
        });
    }

    /// <summary>
    /// HubConnectivity: Pings the Playwright/Selenium Grid hub URL to confirm
    /// it's reachable before grid-dependent tests run.
    ///
    /// WHY CHECK THE HUB: Tests that use a remote grid (Docker Compose or Kubernetes)
    /// will fail immediately with connection errors if the hub isn't running. This
    /// sanity test catches that condition with a clear message before any app tests
    /// start, saving debugging time.
    ///
    /// WHY IGNORE (NOT FAIL) IF HUB:URL IS NOT CONFIGURED: The hub is only needed
    /// when running tests in grid mode. Local development runs don't use a grid.
    /// Using Assert.Ignore() gracefully skips this test rather than failing it when
    /// no hub URL is configured — the skip appears in reports so you know it ran
    /// but wasn't applicable in this environment.
    ///
    /// WHY NOT ALWAYS ASSERT: If the hub is configured but unreachable, the test
    /// fails with "Could not connect to hub" — a clear actionable message.
    /// If the hub is not configured, the test is ignored — also a clear message.
    /// Both outcomes are more helpful than a generic failure.
    /// </summary>
    [Test]
    [Category("sanity")]
    public async Task HubConnectivity()
    {
        // Read the hub URL from config — only present when running in grid mode.
        string hubUrl = ConfigReader.GetProperty("hub:url");

        // If hub:url is not configured, this test isn't applicable in this environment.
        // Assert.Ignore() marks the test as "skipped" in the report — not a failure,
        // not a pass, just "not applicable here."
        if (string.IsNullOrWhiteSpace(hubUrl))
        {
            Assert.Ignore("hub:url is not configured — skipping hub connectivity check.");
            return;
        }

        // Create a short-lived HttpClient for this health check.
        // 'using var' ensures it's disposed (connection closed) after the test,
        // even if an exception is thrown.
        using var http = new System.Net.Http.HttpClient();
        http.Timeout = TimeSpan.FromSeconds(10); // Fail fast — hub should respond in ms if it's up

        try
        {
            // Send a GET request to the hub URL. The Selenium Grid /status endpoint
            // returns a JSON health payload — any non-5xx response means it's alive.
            var response = await http.GetAsync(hubUrl);

            // Any response below 500 (200 OK, 301 redirect, etc.) means the hub
            // is reachable and responding. A 5xx means the hub is up but broken.
            Assert.That((int)response.StatusCode, Is.LessThan(500),
                $"Hub at {hubUrl} should respond with a non-5xx status.");
        }
        catch (Exception ex)
        {
            // Connection refused, DNS failure, timeout — the hub is unreachable.
            Assert.Fail($"Could not connect to hub at {hubUrl}: {ex.Message}");
        }
    }
}
